### PR TITLE
Add new Equatable interface for easier model comparison

### PR DESCRIPTION
**Pull request by stloyd**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| License       | MIT
Replaces #2408.

### CODE REVIEW

**Comment by stloyd** [comment]
@pjedrzejewski @aRn0D @umpirsky feedback? (note that specs are not adjusted yet ;))

**Comment by aRn0D** [comment]
I like it !

**Comment by umpirsky** [comment]
Looks cool. What do you think about some default implementation of `isEqualTo()`, they just look copy pasted?

File: src/Sylius/Component/Addressing/Model/Province.php [hunk 1]
```diff
@@ -112,4 +112,20 @@ public function setCountry(CountryInterface $country = null)
...
+    {
+        if (is_object($data)) {
+            if (!$data instanceof ProvinceInterface) {
+                return false;
+            }
+
+            return $data->getId() === $this->id;
+        }
+
+        return $data === $this->id;
```
Comment (amenophis) [inline src/Sylius/Component/Addressing/Model/Province.php]: Why this ?

File: src/Sylius/Component/Core/Model/User.php [hunk 2]
```diff
@@ -371,4 +371,20 @@ public function getRoles()
...
+    {
+        if (is_object($data)) {
+            if (!$data instanceof UserInterface) {
+                return false;
+            }
+
+            return $data->getId() === $this->id;
+        }
+
+        return $data === $this->id;
```
Comment (amenophis) [inline src/Sylius/Component/Core/Model/User.php]: Why this ?

File: src/Sylius/Component/Addressing/Model/Country.php [hunk 3]
```diff
@@ -160,6 +160,22 @@ public function hasProvince(ProvinceInterface $province)
...
+    {
+        if (is_object($data)) {
+            if (!$data instanceof CountryInterface) {
+                return false;
+            }
+
+            return $data->getId() === $this->id;
+        }
+
+        return $data === $this->id;
```
Comment (amenophis) [inline src/Sylius/Component/Addressing/Model/Country.php]: Why this ?

File: src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php [hunk 4]
```diff
@@ -140,7 +140,7 @@ protected function findOrderOr404($number)
             throw $this->createNotFoundException('The order does not exist.');
         }
 
-        if (!$this->get('security.context')->isGranted('ROLE_SYLIUS_ADMIN') && $this->getUser()->getId() !== $order->getUser()->getId()) {
+        if (!$this->get('security.context')->isGranted('ROLE_SYLIUS_ADMIN') && !$this->getUser()->isEqualTo($order->getUser())) {
```
Comment (amenophis) [inline src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php]: It would be better to use a Voter and use isGranted to check the owner of the Order

Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php]: This is out of scope for this PR, please open new issue for this.

File: src/Sylius/Component/Addressing/Model/Country.php (hunk 3 above)
Comment (stloyd) [inline src/Sylius/Component/Addressing/Model/Country.php]: Cause some stuff saves just ID of the object, i.e. `ShippingCountryRuleChecker`...

**Comment by stloyd** [comment]
@pjedrzejewski Feedback?

**Comment by pjedrzejewski** [comment]
Comparable?

**Comment by stloyd** [comment]
@pjedrzejewski @aRn0D @umpirsky I have drop custom interface & re-used Doctrine one...

**Comment by aRn0D** [comment]
@pjedrzejewski It sounds good ! but @stloyd the build failed!

**Comment by michalmarcinkowski** [comment]
Returning ```1``` when objects are equal is a misuse of the Comparable interface. This interface is used to compare two value objects to each other for similarity and find out whether the objects are equal, less or more.

If we want to compare for equality IMO we should use method 'equals' which will return true or false. 
Instead of:
```
if (1 === $object1->compareTo($object2)) {
    ... do something
}
```
We can make a simpler check then
```
if ($object1->equals($object2)) {
    ... do something
}
```

What do you think guys?

**Comment by aRn0D** [comment]
@michalmarcinkowski This is not false but do we need to create a custom interface or do we use an existing one ?

**Comment by michalmarcinkowski** [comment]
@aRn0D "This is not false" - what do you mean?

I think we will need to create a custom interface, because I don't know any which would provide such method. Unless anyone have an idea?

**Comment by aRn0D** [comment]
@michalmarcinkowski Sorry, I wanted to say "you're right"

**Comment by michalmarcinkowski** [comment]
@aRn0D No problem :)

Unless anyone knows an existing interface, I would suggest to create custom named EquatableInterface or EqualityComparableInterface.

File: src/Sylius/Component/Addressing/Model/Province.php (hunk 1 above)
Comment (stloyd) [inline src/Sylius/Component/Addressing/Model/Province.php]: This is done this way to handle BC when we have just IDs of entities instead of whole entity (think promotions etc.)

File: src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php [hunk 5]
```diff
@@ -140,15 +141,16 @@ protected function findOrderOr404($number)
...
-        ) {
+        if (!$this->get('security.authorization_checker')->isGranted('ROLE_SYLIUS_ADMIN') && !$order->getCustomer()->isEqualTo($this->getCustomer())) {
             throw new AccessDeniedException();
         }
 
         return $order;
     }
 
+    /**
+     * @return null|CustomerInterface
```
Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php]: `CustomerInterface|null`

**Comment by stloyd** [comment]
@pjedrzejewski @michalmarcinkowski @aRn0D @umpirsky I have reworked this PR to introduce new simpler interface instead of depending on Doctrine one, which was confusing I guess.

File: src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php [hunk 6]
```diff
@@ -46,11 +37,9 @@ class ZoneMatcher implements ZoneMatcherInterface
     );
 
     /**
-     * Constructor.
-     *
-     * @param ObjectRepository $repository
+     * @param RepositoryInterface $repository
      */
-    public function __construct(ObjectRepository $repository)
+    public function __construct(RepositoryInterface $repository)
```
Comment (pamil) [inline src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php]: Why?

File: src/Sylius/Component/Addressing/Model/Country.php [hunk 7]
```diff
@@ -146,6 +146,22 @@ public function hasProvince(ProvinceInterface $province)
...
+    {
+        if (is_object($data)) {
+            if (!$data instanceof CountryInterface) {
+                return false;
+            }
+
+            return $data->getIsoName() === $this->isoName || $data->getId() === $this->id;
+        }
+
+        return $data === $this->id;
```
Comment (pamil) [inline src/Sylius/Component/Addressing/Model/Country.php]: Why does number `1` is equal to `Country` instance?

File: src/Sylius/Component/Addressing/Model/ZoneInterface.php [hunk 8]
```diff
@@ -24,7 +24,9 @@
     const TYPE_ZONE = 'zone';
 
     /**
-     * @return mixed
+     * Get ID.
```
Comment (pamil) [inline src/Sylius/Component/Addressing/Model/ZoneInterface.php]: Redundant comment.

File: src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php (hunk 6 above)
Comment (stloyd) [inline src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php]: To be consistent. Check all other parts of the components that use repositories.

File: src/Sylius/Component/Addressing/Model/Country.php (hunk 7 above)
Comment (stloyd) [inline src/Sylius/Component/Addressing/Model/Country.php]: As said before, it's for BC, and you don't compare int to object here, please read code.

File: src/Sylius/Component/Addressing/Model/ZoneInterface.php (hunk 8 above)
Comment (stloyd) [inline src/Sylius/Component/Addressing/Model/ZoneInterface.php]: Rebase issue.

File: src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php (hunk 5 above)
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php]: Rly? This means totally same, and even mine is more correct IMO as `null` is returned faster in context than object...

File: src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Component/Addressing/Matcher/ZoneMatcher.php]: It's not a subject of this PR.

File: src/Sylius/Component/Addressing/Model/Country.php (hunk 7 above)
Comment (pamil) [inline src/Sylius/Component/Addressing/Model/Country.php]: Sure, you don't compare `int` to `object`, but `1` is still equal to `Country`. IMHO that should return false as it's not really equal.

Comment (stloyd) [inline src/Sylius/Component/Addressing/Model/Country.php]: Totally don't get you...
```php
$id = 1;

$country = new Country()
$country->getId(); // let's say it will return 1

$otherCountry = new Country()
$otherCountry->getId(); // let's say it will return 2

// Checkers call `isEqualTo()`
var_dump($promotion->isEliglible($country, ['country' => $id])); // true
var_dump($promotion->isEliglible($country, ['country' => $country])); // true
var_dump($promotion->isEliglible($country, ['country' => 2])); // false
var_dump($promotion->isEliglible($country, ['country' => $otherCountry])); // false

File: src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php (hunk 5 above)
Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Controller/Frontend/Account/OrderController.php]: I mean that we expect to get `CustomerInterface` by this call, getting `null` is a secondary option.

File: src/Sylius/Component/Addressing/Model/Country.php (hunk 7 above)
Comment (pamil) [inline src/Sylius/Component/Addressing/Model/Country.php]: If `PromotionRule` stores ids of resources, it should know how to compare them and use direct comparisons.
